Framework/PythonInterface/mantid/FunctionWrapper.py

```diff
@@ -0,0 +1,480 @@
+       needToFlatten = False
+       for i in range(0, len(self)):
+          if not needToFlatten and isinstance(self[i],CompositeFunctionWrapper):
+             needToFlatten = True
+             
+       if not needToFlatten :
+          return self
+       
+       # Now we know there is a composite function.
+       if isinstance(self,ProductFunctionWrapper):
+          flatSelf = ProductFunctionWrapper()
+       else:
+          flatSelf = CompositeFunctionWrapper()
+          
+       for i in range(0, len(self)):
+          if isinstance(self[i],CompositeFunctionWrapper):
+             currentFunction = self[i].flatten()
+             for j in range(0, len(currentFunction)):
+                flatSelf.fun.add(currentFunction[j].fun)
+          else:
+             flatSelf.fun.add(self[i].fun)
+             
+       return flatSelf          
+ 
+class ProductFunctionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Product Fitting Function
+    """
+    def __init__ (self, *args):
+       """ Called when creating an instance
+           It should not be called directly.
+       """
+       self.pureAddition = False
+       self.pureMultiplication = True
+       return self.initByName("ProductFunction", *args)
+       
+class ConvolutionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Convolution Fitting Function
+    """
+    def __init__ (self, *args):
+       """ Called when creating an instance
+           It should not be called directly.
+       """
+       self.pureAddition = False
+       self.pureMultiplication = False
+       return self.initByName("Convolution", *args)
+       
+class MultiDomainFunctionWrapper(CompositeFunctionWrapper):
+    """ Wrapper class for Product Fitting Function
+    """
+    def __init__ (self, *args, **kwargs):
+       """ Called when creating an instance
+           It should not be called directly
+       """
+       # Assume it's not safe to flatten
+       self.pureAddition = False
+       self.pureMultiplication = False
+       
+       # Create and populate with copied functions
+       self.initByName("MultiDomainFunction", *args)
+               
+       # Tie the global parameters
+       if 'Global' in kwargs:
+          list = kwargs['Global']
+          for name in list:
+             self.tieAll(name)
+       
+       # Set domain indices: 1 to 1       
+       for i in range(0, len(self)):
+          self.fun.setDomainIndex(i, i)
+          
+    def nDomains (self):
+       """ Return number of domains 
+           Used for testing
+       """
+       return self.fun.nDomains()
+     
+        
+def _create_wrapper_function(name):
+    """Create fake functions for the given name
+       It should not be called directly
+                  
+       :param name: name of fake function
+    """
+    # ------------------------------------------------------------------------------------------------
+    def wrapper_function(*args, **kwargs):
```

Comment:
@KarlPalmen registry was a fancy way to say dictionary. Subscription means entering a pair `key:value` in the dictionary whenever a new wrapper class is coded. An automatic way of doing this would be desirable, something like the `DECLARE_FUNCTION` macro.